### DISCUSSION TITLE
Launchpad 2009 is slow turning LEDs on, and LaunchPad MK2 is not reacting

### DISCUSSION THREAD

**Discussion by xnafan** [melanchall/drywetmidi#246] [Q&A]
Hi there
Awesome library! 😊👍👍
I have noticed that my 2009 edition of the Novation Launchpad is slow in responding to commands when I use this library. If I run a loop to turn all LEDs on and the off, I can see them being turned off in a wave, and lighting/turning off the entire pad takes almost an entire second.
Could I be using the library wrong, is the hardware in the 2009 version just slow, or is the MIDI protocol the bottleneck? 🤔 

This is the encapsulating code I'm using: <a href="https://github.com/xnafan/NovationLaunchPadLibrary/blob/master/LaunchPad/NovationLaunchPad.cs">NovationLaunchPadLibrary/NovationLaunchPad.cs</a>[<a href="https://github.com/xnafan/NovationLaunchPadLibrary/blob/master/LaunchPad/NovationLaunchPad.cs" target="_blank" title="New Window">^</a>].
I'm using the IOutputDevice to access the LaunchPad like this: <pre>_outputDevice.SendEvent(new ControlChangeEvent((SevenBitNumber)buttonValue, (SevenBitNumber)0));</pre> in order to turn on a LED.

Another thing. I'm using similar code to try and access the MK2 version of the LaunchPad, but without any luck. I don't even get any events from the IInputDevice.EventReceived event. Are there differences to be aware of between the two models? 

Any help in discovering solutions or just progressing with the problems is much appreciated 😊

Kind regards - Jakob
--
Also posted as a question on the CodeProject page: https://www.codeproject.com/Articles/1275475/DryWetMIDI-Working-with-MIDI-Devices?msg=5934989#xx5934989xx

**Comment by melanchall** [comment]
Hi Jakob,

> If I run a loop to turn all LEDs on and the off, I can see them being turned off in a wave, and lighting/turning off the entire pad takes almost an entire second.
> Could I be using the library wrong, is the hardware in the 2009 version just slow, or is the MIDI protocol the bottleneck?

First of all, it should not be related to the DryWetMIDI unless you do things that you've not shown here. Can you please show the entire loop to send MIDI events?

`SendEvent` just passes an event to MIDI driver and here the area of the library responsibility ends. There are a lot of possible places where the issue can occur. Some of them are MIDI subsystem of OS, MIDI device driver or internal processing of specific MIDI events inside the device.

You can test the speed of sending via some virtual MIDI cable, for example.

> Another thing. I'm using similar code to try and access the MK2 version of the LaunchPad, but without any luck. I don't even get any events from the IInputDevice.EventReceived event.

Please show your code how you setup `IInputDevice` to receive events. Have you `StartEventsListening` call?

Thanks,
Max

**Reply by xnafan** [reply]
Hi Max 😊
Thanks for getting back to me 👍
Here is the minimal code I'm effectively using:
```
IInputDevice inputDevice = InputDevice.GetByName("LaunchPad");
inputDevice.EventReceived += OnEventReceived;
inputDevice.StartEventsListening();
```

The entire code can be seen in the GitHub project I linked to in my original question.

This is the method I've created to turn on a button:
(note that I have created a double array to store the current state of the buttons, to avoid sending "turn on" signals to buttons that are already lit up).
```
public void ButtonOn(int col, int row, ButtonColor color, bool force = false)
    {
        //to save time, don't send a new "turn on" signal, if the light is already on
        if (!force && _boardState.GetButtonColor(col, row).Equals(color)) { return; }
        int buttonValue = 0;
        if (row == 0)
        {
            buttonValue = 104 + col;
            _outputDevice.SendEvent(new ControlChangeEvent((SevenBitNumber)buttonValue, (SevenBitNumber)(byte)color));
        }
        else
        {
            buttonValue = 16 * (row - 1) + col;
            _outputDevice.SendEvent(new NoteOnEvent((SevenBitNumber)buttonValue, (SevenBitNumber)(byte)color));
        }
        _boardState.ChangeButtonState(col, row, color);
    }
```

And this is the code I use to turn all on:

```
 public void AllOn(ButtonColor color, bool force = false)
    {
        for (int x = 0; x < 9; x++)
        {
            for (int y = 0; y < 9; y++)
            {
                ButtonOn(x, y, color, force);
            }
        }
    }
```
Thanks again for taking the time to look it over 😊👍

Kind regards - Jakob

**Reply by melanchall** [reply]
Hi Jakob,

Well, I doesn't see any issues with your loop. My assumption is that's how your device responds to MIDI events. You can even measure how long the loop executing with `Stopwatch`. I'm pretty sure it takes milliseconds.

As for no events received... Hmm, the code is fine. Is there a chance you have several devices plugged with the same name? What devices are shown if you call `InputDevice.GetAll`?

You can also collect all input devices ans subscribe to `EventReceived` event of all of them to see if MIDI events are received from one of the devices or not. Also you can check events receiving with some DAW for example.

Thanks,
Max

**Reply by xnafan** [reply]
Hi Max 😊
Thanks a bundle for looking over my code.
I'm guessing that the 2009 model just might be slow to respond...? 🤔
Regarding the MK2 - is there a different command I have to send in order to subscribe to the buttons and make them light up? 

Kind regards - Jakob

**Reply by melanchall** [reply]
Hi Jakob,

First of all, you probably didn't notice my questions and suggestions:

1. Is there a chance you have several devices plugged with the same name? Maybe MK2 and 2009 have the same name in Windows MIDI subsystem. `InputDevice.GetByName` returns a **FIRST** device with the specified name.
2. What devices are shown if you call `InputDevice.GetAll`? Show me please names of all devices returned by the method when you are trying to receive data from MK2.
3. Check MK2 events receiving with some DAW (for example, Cubase or Logic or whatever).
4. You can even measure how long the loop executing with `Stopwatch` for sending events to 2009:

    ```csharp
    public void AllOn(ButtonColor color, bool force = false)
    {
        var sw = Stopwatch.SrartNew();

        for (int x = 0; x < 9; x++)
        {
            for (int y = 0; y < 9; y++)
            {
                ButtonOn(x, y, color, force);
            }
        }

        sw.Stop();
        var elapsed = sw.Elapsed;
    }
    ```

    Just set breakpoint on `var elapsed = sw.Elapsed;` and show what value is in `elapsed` variable in debug.

As for you last questions.

> I'm guessing that the 2009 model just might be slow to respond...?

I think so.

> Regarding the MK2 - is there a different command I have to send in order to subscribe to the buttons and make them light up?

Unfortunately I can help you with DryWetMIDI only. I have my main job and I work on DryWetMIDI in my spare time, so I just want to take a rest sometimes :) So I didn't know how to manage your device, you should read the manual for the device. If it's managed by MIDI events, well, I can help you with sending events to a MIDI device via my library. But it seems you already know how to do this.

Thanks,
Max

**Comment by xnafan** [comment]
Hi Max 😊
Thanks again for your time and your suggestions 👍
I kept fiddling around, found some lower level documentation and also learned, that there is a "Session" mode on the launchpad MK2, which must be enabled in order to receive events. So I am getting closer to a solution.

Is there any way to buy you a cup of coffee or a beer? Your PayPal doesn't seem to be working (I get the message "This recipient can't accept payments right now." when I try to send you something 😕

Kind regards - Jakob

**Reply by melanchall** [reply]
Hi Jakob,

Good news! :)

> Is there any way to buy you a cup of coffee or a beer?

Unfortunately no. I'm in Russia. Maybe you know, all big payments systems don't work with us now due to the situation that we all know :(

So thank you for using the library and for kind words!

Max
